client/asset/eth/multirpc.go

```diff
@@ -0,0 +1,1368 @@
+		}
+		p.setTip(hdr)
+		providers = append(providers, p)
+
+		// Start websocket listen loop.
+		if sub != nil {
+			go p.subscribeHeaders(ctx, sub, h, log)
+		}
+	}
+
+	if len(providers) == 0 {
+		return nil, fmt.Errorf("failed to connect")
+	}
+
+	log.Debugf("Connected with %d of %d RPC servers", len(providers), len(endpoints))
+
+	success = true
+	return providers, nil
+}
+
+func (m *multiRPCClient) connect(ctx context.Context) (err error) {
+	providers, err := connectProviders(ctx, m.endpoints, m.log, m.chainID)
+	if err != nil {
+		return err
+	}
+
+	m.providerMtx.Lock()
+	m.providers = providers
+	m.providerMtx.Unlock()
+
+	var connections int
+	for _, p := range m.providerList() {
+		if _, err := p.bestHeader(ctx, m.log); err != nil {
+			m.log.Errorf("Failed to synchrnoize header from %s: %v", p.host, err)
+		} else {
+			connections++
+		}
+	}
+	// TODO: Require at least two if all connections are non-local.
+	if connections == 0 {
+		return fmt.Errorf("no connections established")
+	}
+
+	go func() {
+		<-ctx.Done()
+		for _, p := range m.providerList() {
+			p.ec.Close()
+		}
+	}()
+
+	return nil
+}
+
+func (m *multiRPCClient) reconfigure(ctx context.Context, settings map[string]string) error {
+	providerDef := settings[providersKey]
+	if len(providerDef) == 0 {
+		return errors.New("no providers specified")
+	}
+	endpoints := strings.Split(providerDef, " ")
+	providers, err := connectProviders(ctx, endpoints, m.log, m.chainID)
+	if err != nil {
+		return err
+	}
+	m.providerMtx.Lock()
+	oldProviders := m.providers
+	m.providers = providers
+	m.endpoints = endpoints
+	m.providerMtx.Unlock()
+	for _, p := range oldProviders {
+		p.ec.Close()
+	}
+	return nil
+}
+
+func (m *multiRPCClient) cachedReceipt(txHash common.Hash) *types.Receipt {
+	m.receipts.Lock()
+	defer m.receipts.Unlock()
+
+	cached := m.receipts.cache[txHash]
+
+	// Periodically clean up the receipts.
+	if time.Since(m.receipts.lastClean) > time.Minute*20 {
+		m.receipts.lastClean = time.Now()
+		defer func() {
+			for txHash, rec := range m.receipts.cache {
+				if time.Since(rec.lastAccess) > receiptCacheExpiration {
+					delete(m.receipts.cache, txHash)
+				}
+			}
+		}()
+	}
+
+	// If confirmed or if it was just fetched, return it as is.
+	if cached != nil {
+		// If the cached receipt has the requisite confirmations, it's always
+		// considered good and we'll just update the lastAccess stamp so we don't
+		// delete it from the map.
+		// If it's not confirmed, we never update the lastAccess stamp, which just
+		// serves to age out the receipt so a new one can be requested and
+		// confirmations checked again.
+		if cached.confirmed {
+			cached.lastAccess = time.Now()
+		}
+		if time.Since(cached.lastAccess) < unconfirmedReceiptExpiration {
+			return cached.r
+		}
+	}
+	return nil
+}
+
+// cleanReceipts cleans up the receipt cache, deleting any receipts that haven't
+// been access for > receiptCacheExpiration.
+func (m *multiRPCClient) cleanReceipts() {
```

Comment:
`cleanReceipts` isn't called anywhere.  Looks like you merged it into `cachedReceipt` and this method can be removed.